### PR TITLE
TM4J-1449 add support for TM4J Cloud

### PR DESCRIPTION
**Pull request by dragosmc**

### CODE REVIEW

File: src/main/java/com/adaptavist/tm4j/jenkins/extensions/Instance.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
+package com.adaptavist.tm4j.jenkins.extensions;
+
+import com.mashape.unirest.http.HttpResponse;
+import com.mashape.unirest.http.JsonNode;
+import com.mashape.unirest.http.exceptions.UnirestException;
+
+import java.io.File;
+
+public interface Instance {
```
Comment (otaviomedeiros) [inline src/main/java/com/adaptavist/tm4j/jenkins/extensions/Instance.java]: Don't you think it should be called `JiraInstance`?

File: src/main/java/com/adaptavist/tm4j/jenkins/extensions/JiraInstance.java [hunk 2]
```diff
@@ -7,12 +7,11 @@
 import hudson.util.Secret;
 import org.apache.http.client.HttpClient;
 import org.apache.http.impl.client.HttpClientBuilder;
-import org.kohsuke.stapler.DataBoundConstructor;
 
 import java.io.File;
 import java.text.MessageFormat;
 
-public class JiraInstance {
+public class JiraInstance implements Instance {
```
Comment (otaviomedeiros) [inline src/main/java/com/adaptavist/tm4j/jenkins/extensions/JiraInstance.java]: I'd call it `JiraServerInstance`

File: src/main/java/com/adaptavist/tm4j/jenkins/extensions/buildsteps/FeatureFilesDownloader.java [hunk 3]
```diff
@@ -45,25 +43,24 @@ public FeatureFilesDownloader(String serverAddress, String projectKey, String ta
...
-        List<JiraInstance> jiraInstances = getDescriptor().getJiraInstances();
+        List<Instance> jiraInstances = getDescriptor().getJiraInstances();
         try {
             new Validator().validateProjectKey(this.projectKey)
                     .validateTargetPath(this.targetPath)
                     .validateServerAddress(this.serverAddress);
             String tql = format("testCase.projectKey = '%s'", this.projectKey);
             Tm4jJiraRestClient tm4jJiraRestClient = new Tm4jJiraRestClient(jiraInstances, serverAddress);
-            tm4jJiraRestClient.exportFeatureFiles(run.getRootDir(), workspace, targetPath, tql, logger);
+            tm4jJiraRestClient.importFeatureFiles(run.getRootDir(), workspace, targetPath, tql, logger);
```
Comment (otaviomedeiros) [inline src/main/java/com/adaptavist/tm4j/jenkins/extensions/buildsteps/FeatureFilesDownloader.java]: Nice! `tm4jJiraRestClient` is actually importing feature files.

File: src/main/java/com/adaptavist/tm4j/jenkins/http/Tm4jJiraRestClient.java [hunk 4]
```diff
@@ -24,40 +25,40 @@
...
             logger.printf("%s The generated ZIP file couldn't be deleted. Please check folder permissions and delete the file manually: " + file.getAbsolutePath() + " %n", INFO);
         }
     }
 
     public void uploadCustomFormatFile(String directory, String filePath, String projectKey, Boolean autoCreateTestCases, final PrintStream logger) throws Exception {
         File file = new FileReader().getZip(directory, filePath);
-        HttpResponse<JsonNode> jsonResponse = jiraInstance.importCustomFormatBuildResult(projectKey, autoCreateTestCases, file);
-        processImportingResultsResponse(jsonResponse, logger);
-        if(!file.delete()) {
+        HttpResponse<JsonNode> jsonResponse = jiraInstance.exportCustomFormatBuildResult(projectKey, autoCreateTestCases, file);
```
Comment (otaviomedeiros) [inline src/main/java/com/adaptavist/tm4j/jenkins/http/Tm4jJiraRestClient.java]: In this case i'm confused, `jiraInstance` is importing from or exporting from?

File: src/main/java/com/adaptavist/tm4j/jenkins/extensions/JiraInstance.java (hunk 2 above)
Comment (fnmunhoz) [reply src/main/java/com/adaptavist/tm4j/jenkins/extensions/JiraInstance.java]: Is this going to change the persisted XML file if we rename the class? I don't really know. Just wondering.

File: src/main/java/com/adaptavist/tm4j/jenkins/extensions/Instance.java (hunk 1 above)
Comment (dragosmc) [reply src/main/java/com/adaptavist/tm4j/jenkins/extensions/Instance.java]: Yes, I'd make more sense to be called `JiraInstance` and have `JiraServerInstance` and `JiraCloudInstance`, but because Jenkins saves the configuration in an XML file looking like 

`<com.adaptavist.tm4j.jenkins.extensions.configuration.Tm4jGlobalConfiguration plugin="tm4j-automation@1.0.0">
  <jiraInstances>
    <com.adaptavist.tm4j.jenkins.extensions.JiraInstance>
      <serverAddress>http://localhost:2990/jira</serverAddress>
      <username>admin</username>
      <password>admin</password>
    </com.adaptavist.tm4j.jenkins.extensions.JiraInstance>
  </jiraInstances>
</com.adaptavist.tm4j.jenkins.extensions.configuration.Tm4jGlobalConfiguration>`

That will then get deserialised in the specific beans. Renaming the `JiraInstance`, which was used as the bean for server, into `JiraServerInstance` will invalidate all previous configurations.

Currently, the XML looks like
`<com.adaptavist.tm4j.jenkins.extensions.configuration.Tm4jGlobalConfiguration plugin="tm4j-automation@1.1.0">
  <jiraInstances>
    <com.adaptavist.tm4j.jenkins.extensions.JiraCloudInstance>
      <jwt>token</jwt>
      <name>URL</name>
    </com.adaptavist.tm4j.jenkins.extensions.JiraCloudInstance>
    <com.adaptavist.tm4j.jenkins.extensions.JiraInstance>
      <serverAddress>URL</serverAddress>
      <username>admin</username>
      <password>admin</password>
    </com.adaptavist.tm4j.jenkins.extensions.JiraInstance>
  </jiraInstances>
</com.adaptavist.tm4j.jenkins.extensions.configuration.Tm4jGlobalConfiguration>`

File: src/main/java/com/adaptavist/tm4j/jenkins/extensions/JiraInstance.java (hunk 2 above)
Comment (dragosmc) [reply src/main/java/com/adaptavist/tm4j/jenkins/extensions/JiraInstance.java]: Please see my previous comment :)

File: src/main/java/com/adaptavist/tm4j/jenkins/http/Tm4jJiraRestClient.java (hunk 4 above)
Comment (dragosmc) [reply src/main/java/com/adaptavist/tm4j/jenkins/http/Tm4jJiraRestClient.java]: Usually the concept is `import from` and `export to` - so the plugin exports the custom/cucumber results and imports the feature files. I think I've renamed everything in the plugin based on this assumption which, imho, makes more sense when reading the code.

File: README.md [hunk 5]
```diff
@@ -1,10 +1,10 @@
 # TM4J Jenkins Plugin
 
-A plugin for Jenkins built for integrating automated tests with [Test Management for Jira Server](https://marketplace.atlassian.com/apps/1213259/test-management-for-jira?hosting=server&tab=overview) (TM4J).  
+A plugin for Jenkins built for integrating automated tests with [Test Management for Jira Server](https://marketplace.atlassian.com/apps/1213259/test-management-for-jira?hosting=server&tab=overview) and [Test Management for Jira Cloud](https://marketplace.atlassian.com/apps/1213259/tm4j-test-management-for-jira?tab=overview&hosting=cloud) (TM4J).  
 
 The plugin introduces two new configuration steps for Jenkins jobs:
-1) A build step to export Gherkin feature files (BDD) from TM4J
-1) A post-build step to import the test results back to TM4J
+1) A build step to import Gherkin feature files (BDD) from TM4J (currently available only for Test Management for Jira Server)
+1) A post-build step to export the test results back to TM4J
```
Comment (pelizza) [inline README.md]: What if we use the same naming convention as we have for the Build Steps title: "download Gherkin..." and "publish test results..."

File: README.md [hunk 6]
```diff
@@ -1,10 +1,10 @@
 # TM4J Jenkins Plugin
 
-A plugin for Jenkins built for integrating automated tests with [Test Management for Jira Server](https://marketplace.atlassian.com/apps/1213259/test-management-for-jira?hosting=server&tab=overview) (TM4J).  
+A plugin for Jenkins built for integrating automated tests with [Test Management for Jira Server](https://marketplace.atlassian.com/apps/1213259/test-management-for-jira?hosting=server&tab=overview) and [Test Management for Jira Cloud](https://marketplace.atlassian.com/apps/1213259/tm4j-test-management-for-jira?tab=overview&hosting=cloud) (TM4J).
```
Comment (pelizza) [inline README.md]: The last "(TM4J)" is a bit lost there. Suggestion:

A plugin for Jenkins built for integrating automated tests with Test Management for Jira (TM4J). Compatible with both Jira Server and Jira Cloud. For more information, please check [TM4J Server](https://marketplace.atlassian.com/apps/1213259/test-management-for-jira?hosting=server&tab=overview) and [TM4J Cloud](https://marketplace.atlassian.com/apps/1213259/tm4j-test-management-for-jira?tab=overview&hosting=cloud).

File: pom.xml [hunk 7]
```diff
@@ -39,7 +39,7 @@
     </scm>
 
     <properties>
-        <jenkins.version>2.73.3</jenkins.version>
+        <jenkins.version>2.164.1</jenkins.version>
```
Comment (pelizza) [inline pom.xml]: I remember there was a reason to set this specific number here. This version means the first Jenkins version we are compatible with.
@chicofaccin any thoughts?

File: src/main/java/com/adaptavist/tm4j/jenkins/extensions/buildsteps/FeatureFilesDownloader.java (hunk 3 above)
Comment (pelizza) [reply src/main/java/com/adaptavist/tm4j/jenkins/extensions/buildsteps/FeatureFilesDownloader.java]: As per my previous comment, what if we stick to the naming convention we have for the build steps title? "downloadFeatureFiles". I think that would be clearer.

File: src/main/java/com/adaptavist/tm4j/jenkins/http/Tm4jJiraRestClient.java (hunk 4 above)
Comment (pelizza) [reply src/main/java/com/adaptavist/tm4j/jenkins/http/Tm4jJiraRestClient.java]: As per my previous comments, what if we stick to the naming convention we have for the build steps title? "publishCustomFormatBuildResult". I think that would be clearer.

File: src/main/resources/com/adaptavist/tm4j/jenkins/extensions/configuration/Tm4jGlobalConfiguration/config.jelly [hunk 8]
```diff
@@ -2,30 +2,37 @@
 <j:jelly xmlns:j="jelly:core"
          xmlns:f="/lib/form">
     <f:section title="Test Management for Jira">
-
-        <f:entry title="Jira server(s)">
-            <f:repeatable var="tm4jInstancesList" name="jiraInstances" items="${instance.jiraInstances}"
+        <f:entry title="Jira instance(s)">
+            <f:repeatable varStatus="status" var="tm4jInstancesList" name="jiraInstances" items="${instance.jiraInstances}"
                           noAddButton="false" minimum="0">
                 <table>
```
Comment (pelizza) [inline src/main/resources/com/adaptavist/tm4j/jenkins/extensions/configuration/Tm4jGlobalConfiguration/config.jelly]: I think would be nice to add a link to the documentation of each product (Server/Cloud) somewhere here. Please check out with our POs what they think? May be a separate PR/task though...

Comment (dragosmc) [reply src/main/resources/com/adaptavist/tm4j/jenkins/extensions/configuration/Tm4jGlobalConfiguration/config.jelly]: Each help section has the link to the Wiki page - I wouldn't add another link in there.

Comment (pelizza) [reply src/main/resources/com/adaptavist/tm4j/jenkins/extensions/configuration/Tm4jGlobalConfiguration/config.jelly]: Oh, awesome, I didn't know that! I suppose by wiki page you mean doco...

File: src/main/resources/com/adaptavist/tm4j/jenkins/extensions/configuration/Tm4jGlobalConfiguration/help-jwt.html [hunk 9]
```diff
@@ -0,0 +1,3 @@
+<div>
```
Comment (pelizza) [inline src/main/resources/com/adaptavist/tm4j/jenkins/extensions/configuration/Tm4jGlobalConfiguration/help-jwt.html]: I just realised the help for the URL field for Jira Server is outdated now that we have Jira Cloud support:
Screenshot 2019-04-05 at 11 00 42

Would be nice if you could update it and use a similar template with the link to our documentation.

File: src/main/resources/com/adaptavist/tm4j/jenkins/extensions/configuration/Tm4jGlobalConfiguration/config.jelly (hunk 8 above)
Comment (dragosmc) [reply src/main/resources/com/adaptavist/tm4j/jenkins/extensions/configuration/Tm4jGlobalConfiguration/config.jelly]: No, I mean the Jenkins wiki

File: src/main/java/com/adaptavist/tm4j/jenkins/extensions/buildsteps/FeatureFilesDownloader.java [hunk 10]
```diff
@@ -45,25 +43,24 @@ public FeatureFilesDownloader(String serverAddress, String projectKey, String ta
...
         } catch (NoTestCasesFoundException e) {
             logger.printf("%s No feature files have been found for project " +  this.projectKey + ". %n", ERROR);
             run.setResult(Result.FAILURE);
             throw new RuntimeException();
         } catch (Exception e) {
             run.setResult(Result.FAILURE);
             logger.printf("%s There was an error while trying to download feature files from Test Management for Jira. Error details: %n", ERROR);
-            logger.printf(ERROR);
```
Comment (pelizza) [inline src/main/java/com/adaptavist/tm4j/jenkins/extensions/buildsteps/FeatureFilesDownloader.java]: Removed by purpose?

Comment (dragosmc) [reply src/main/java/com/adaptavist/tm4j/jenkins/extensions/buildsteps/FeatureFilesDownloader.java]: Yes, we print it in the line above.

File: src/main/java/com/adaptavist/tm4j/jenkins/extensions/Instance.java (hunk 1 above)
Comment (otaviomedeiros) [reply src/main/java/com/adaptavist/tm4j/jenkins/extensions/Instance.java]: Very well observed!

File: src/main/java/com/adaptavist/tm4j/jenkins/http/Tm4jJiraRestClient.java (hunk 4 above)
Comment (otaviomedeiros) [reply src/main/java/com/adaptavist/tm4j/jenkins/http/Tm4jJiraRestClient.java]: That's the point @dragosmc. `tm4jJiraRestClient.exportCustomFormatBuildResult ` would make sense for me but `jiraInstance.exportCustomFormatBuildResult` doesn't make sense because jira doesn't export anything, it imports. Anyway, free interpretation here :)
